### PR TITLE
mango: add $beginsWith operator

### PR DESCRIPTION
**Pull request by willholley**
## Overview

Adds a `$beginsWith` operator to selectors, with json and text index support. This is a compliment / precursor to optimising `$regex` support as proposed in https://github.com/apache/couchdb/pull/4776.

For `json` indexes, a $beginsWith operator translates into a key range query, as is common practice for _view queries. For example, to find all rows with a key beginning with "W", we can use a range `start_key="W", end_key="W\ufff0"`. Given Mango uses compound keys, this is slightly more complex in practice, but the idea is the same. As with other range operators (`$gt`, `$gte`, etc), `$beginsWith` can be used in combination with equality operators and result sorting but must result in a contiguous key range. That is, a range of `start_key=[10, "W"], end_key=[10, "W\ufff0", {}]` would be valid, but `start_key=["W", 10], end_key=["W\ufff0", 10, {}]` would not, because the second element of the key may result in a non-contiguous range.

For text indexes, `$beginsWith` translates to a Lucene query on the specified field of `W*`.

If a non-string operand is provided to `$beginsWith`, the request will fail with a 400 / `invalid_operator` error.

## Testing recommendations

I've only tested this with json indexes so far because the dev container doesn't have `text` index support.

A basic eunit test is added for the new `selector` support. You can run these using `make eunit apps=mango`.

Integration tests are added to the Mango test suite, which can be run via `make mango-test`.

## Related Issues or Pull Requests

n/a

## Checklist

- [ ] Code is written and works correctly
- [ ] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] Documentation changes were made in the `src/docs` folder
- [ ] Documentation changes were backported (separated PR) to affected branches

### CODE REVIEW

File: src/mango/src/mango_selector.erl [hunk 1]
```diff
@@ -1054,4 +1061,29 @@ fields_nor_test() ->
...
     ?assertEqual([<<"field1">>, <<"field2">>], fields_of(Selector2)).
 
+match_beginswith_test() ->
+    Doc =
+        {[
+            {<<"_id">>, <<"foo">>},
+            {<<"_rev">>, <<"bar">>},
+            {<<"user_id">>, 11}
+        ]},
+    Check = fun(Field, Prefix) ->
```
Comment (pgj) [inline src/mango/src/mango_selector.erl:1071]: Instead of copying, maybe it would be worth to lift the `Check` function to the module level and hook it up from here and `match_demo_test_/0` above (i.e. sharing).  This test case could be moved closed to that too.

File: src/mango/src/mango_selector_text.erl [hunk 2]
```diff
@@ -142,6 +142,9 @@ convert(Path, {[{<<"$exists">>, ShouldExist}]}) ->
         true -> FieldExists;
         false -> {op_not, {FieldExists, false}}
     end;
+convert(Path, {[{<<"$beginsWith">>, Arg}]}) ->
```
Comment (pgj) [inline src/mango/src/mango_selector_text.erl:145]: Please add a unit test for this branch.

File: src/mango/test/25-beginswith-test.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    def test_basic(self):
+        docs = self.db.find({"location": {"$beginsWith": "A"}})
+
+        self.assertEqual(len(docs), 2)
+        self.assertDocIds(["aaa", "abc"], docs)
+
+    def test_json_range(self):
+        explain = self.db.find({"location": {"$beginsWith": "A"}}, explain=True)
+        self.assertEqual(explain["mrargs"]["start_key"], ["A"])
```
Comment (pgj) [inline src/mango/test/25-beginswith-test.py:50]: `explain["mrargs"]` could be just tagged with a name, e.g. `mrargs`, so that later it is enough to write `mrargs["start_key"]` and `mrargs["end_key"]`.

File: src/mango/test/25-beginswith-test.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        self.assertEqual(explain["mrargs"]["start_key"], ["Eddie", "A"])
+        end_key_bytes = to_utf8_bytes(explain["mrargs"]["end_key"])
+        self.assertEqual(end_key_bytes, [b"Eddie", b"A\xef\xbf\xbd", b"<MAX>"])
+
+        docs = self.db.find(selector)
+        self.assertEqual(len(docs), 1)
+        self.assertDocIds(["abc"], docs)
+
+    def test_sort_asc(self):
```
Comment (pgj) [inline src/mango/test/25-beginswith-test.py:66]: The sort tests could be merged into a single case with `self.subTest`.  This way it becomes possible to share the commonalities and parametrize them with the differences.  With subtests, it will still be possible to identify each case.  Examples of that could be seen in  `01-index-crud-test` and `23-explain-indexes-test`.

File: src/mango/test/25-beginswith-test.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        try:
+            self.db.find({"_id": {"$beginsWith": True}})
+        except Exception as e:
+            self.assertEqual(e.response.status_code, 400)
+            resp = e.response.json()
+            self.assertEqual(resp["error"], "invalid_operator")
+        else:
+            raise AssertionError("expected find error")
+
+    def test_does_not_match_non_string_value(self):
```
Comment (pgj) [inline src/mango/test/25-beginswith-test.py:108]: There could also be a similar test case when there are no matches for a string field?

**Comment by pgj** [comment]
Great work!  I think it would be worth to cover the `text` indexes with tests in this PR otherwise it can easily go forgotten.

File: src/mango/src/mango_selector.erl (hunk 1 above)
Comment (willholley) [reply src/mango/src/mango_selector.erl:1071]: the `Check` function is test-specific so I'm not convinced it makes sense to share it. I don't think `match_demo_test_` is actually running any tests (this confused me for a while but note it calls `?_assertEqual` instead of `?assertEqual` which appears to be a no-op.

Comment (nickva) [reply src/mango/src/mango_selector.erl:1071]: The eunit test suite vs test vs generated test (a fun) is quite confusing. It's usually simpler to use basic test functions. For a suite, the setup/teardown with `?TDEF/?TDEF_FE` can work well.

Since we have two tests, it could be simpler to have a helper check function like @pgj suggested. The Doc it could be defined in a macro like ?TEST_DOC or re-defined inside the check match helper function, whichever looks better. Then, tests can just be a list of simple asserts:

```erlang
match_demo_test() ->
   ?assert(check_match({[{<<"user_id">>, 11}]} , ?TEST_DOC))),
   ?assertNot(check_match({[{<<"user_id">>, 1234}]}, ?TEST_DOC))),
   ...
```

So it looks like a simple function. If it fails on an assert we should get the exact line number and the expression where it fails.

File: src/mango/test/25-beginswith-test.py (hunk 4 above)
Comment (willholley) [reply src/mango/test/25-beginswith-test.py:66]: although the sort asc/desc tests look structurally similar, there isn't really much commonality. It's subtle, but notice all the assertions are different so would need to be parameterized. Given this, I don't think there's much to be gained by refactoring with `self.subTest`.

File: src/mango/test/25-beginswith-test.py (hunk 5 above)
Comment (willholley) [reply src/mango/test/25-beginswith-test.py:115]: good thought - done

File: src/mango/src/mango_selector.erl (hunk 1 above)
Comment (pgj) [reply src/mango/src/mango_selector.erl:1071]: @willholley ah, I see -- I got probably mislead by the comment that was copied over from the other similarly-looking definition.  On the other hand, I tend to agree with @nickva regarding the actual definition itself.  Embedding the use of `Doc` variable internally makes the test cases a bit less explicit and hard to follow.  I am aware that this is coming from the other test case but perhaps we could cha(lle)nge the original structure here.

With regard to the no-op test case, maybe this is the time to either fix or scratch it?

File: src/mango/test/25-beginswith-test.py (hunk 3 above)
Comment (pgj) [reply src/mango/test/25-beginswith-test.py]: @willholley are you planning to modify the code according this suggestion?  I am only asking because I saw that you sent a "thumbs up" reaction but there was no associated change.

Comment (pgj) [reply src/mango/test/25-beginswith-test.py]: Oop, sorry.  For some reason, I was looking at some old version of the code.

File: src/mango/test/25-beginswith-test.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        self.assertEqual(mrargs["start_key"], ["A"])
+        end_key_bytes = to_utf8_bytes(mrargs["end_key"])
+        self.assertEqual(end_key_bytes, [b"A\xef\xbf\xbd", b"<MAX>"])
+        self.assertEqual(mrargs["direction"], "fwd")
+
+    def test_sort_desc(self):
+        selector = {"location": {"$beginsWith": "A"}}
+        mrargs = self.get_mrargs(selector, sort=[{"location": "desc"}])
+
+        start_key_bytes = to_utf8_bytes(mrargs["end_key"])
```
Comment (pgj) [inline src/mango/test/25-beginswith-test.py:84]: I think this is not right: either the variable name (`start_key_bytes`) or the corresponding field name (`"end_key"`) is wrong here.

**Comment by pgj** [comment]
I miss the parts related to the documentation.

File: src/mango/test/25-beginswith-test.py (hunk 4 above)
Comment (pgj) [reply src/mango/test/25-beginswith-test.py]: The proposed formulation helps to emphasize the differences, suggests to treat the test cases as data and makes it easier to see the logical connections between on the higher level.  For what it is worth, rewriting the code in this way already helped me to identify a mistake:

```python
    def test_sort(self):
        selector = {"location": {"$beginsWith": "A"}}
        cases = [
            {
                "sort": ["location"],
                "start_key": [b"A"],
                "end_key": [b"A\xef\xbf\xbd", b"<MAX>"],
                "direction": "fwd",
            },
            {
                "sort": [{"location": "desc"}],
                "start_key": [b"A\xef\xbf\xbd", b"<MAX>"],
                "end_key": [b"A"],
                "direction": "rev",
            },
        ]

        for case in cases:
            with self.subTest(sort=case["sort"]):
                mrargs = self.get_mrargs(selector, sort=case["sort"])
                self.assertEqual(to_utf8_bytes(mrargs["start_key"]), case["start_key"])
                self.assertEqual(to_utf8_bytes(mrargs["end_key"]), case["end_key"])
                self.assertEqual(mrargs["direction"], case["direction"])
```

**Comment by pgj** [comment]
Please add `text` tests for the integration test suite.  I am fine if they do not fully work (like a draft) I can test them and provide feedback.  For your information, I am currently working on introducing support for testing with Clouseau in the PRs -- you can also wait for that.

File: src/mango/test/25-beginswith-test.py (hunk 4 above)
Comment (willholley) [reply src/mango/test/25-beginswith-test.py]: okay you've convinced me ;)

File: src/mango/src/mango_selector.erl (hunk 1 above)
Comment (willholley) [reply src/mango/src/mango_selector.erl:1071]: Ah I see now that it's not a no-op but a generated test. I'll try a version using the basic test functions as suggested.

**Comment by willholley** [comment]
> Please add `text` tests for the integration test suite. I am fine if they do not fully work (like a draft) I can test them and provide feedback. For your information, I am currently working on introducing support for testing with Clouseau in the PRs -- you can also wait for that.

It would certainly be helpful to have Clouseau in the CI and, presumably, then in the dev container which uses the same CI images.

**Comment by willholley** [comment]
> Please add `text` tests for the integration test suite. I am fine if they do not fully work (like a draft) I can test them and provide feedback. For your information, I am currently working on introducing support for testing with Clouseau in the PRs -- you can also wait for that.

The `03-operator-test.py` updates include integration tests for text indexes that check the result is as expected. I didn't add integration tests that check the args passed to Lucene, but these are covered by eunit tests (it's much simpler than the mrargs parameters which involve a range scan).

**Comment by pgj** [comment]
Okay.  Right now there is a failing test case (for the `text` indexes):

```console
======================================================================
FAIL: test_beginswith (03-operator-test.OperatorTextTests.test_beginswith)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/gaborpali/projects/github/couchdb/src/mango/test/03-operator-test.py", line 146, in test_beginswith
    self.assertEqual(len(docs), 2)
AssertionError: 0 != 2

----------------------------------------------------------------------
Ran 384 tests in 37.973s

FAILED (failures=1, skipped=1)
```

File: src/mango/src/mango_selector_text.erl [hunk 7]
```diff
@@ -142,6 +142,10 @@ convert(Path, {[{<<"$exists">>, ShouldExist}]}) ->
         true -> FieldExists;
         false -> {op_not, {FieldExists, false}}
     end;
+convert(Path, {[{<<"$beginsWith">>, Arg}]}) when is_binary(Arg) ->
+    Suffix = <<"*">>,
+    PrefixSearch = value_str(<<Arg/binary, Suffix/binary>>),
+    {op_field, {make_field(Path, Arg), PrefixSearch}};
```
[the remark this answers is not in the document]
Comment (pgj) [reply src/mango/src/mango_selector_text.erl:148]: It is almost there.  The problem with this version is that it will quote the `*` (wildcard) therefore neutralize that.  Here is an alternative that fixed `03-operator-test.OperatorTextTests.test_beginswith` for me:

```erlang
convert(Path, {[{<<"$beginsWith">>, Arg}]}) when is_binary(Arg) ->
    Prefix = mango_util:lucene_escape_query_value(Arg),
    Suffix = <<"*">>,
    PrefixSearch = <<Prefix/binary, Suffix/binary>>,
    {op_field, {make_field(Path, Arg), PrefixSearch}};
```

[the remark this answers is not in the document]
Comment (pgj) [reply src/mango/src/mango_selector_text.erl:148]: On that note, I would recommend to include spaces or other special characters (subject to escaping) in the strings used in the tests.  For example (note the space after `"New"`):

```python
         def test_beginswith(self):
             docs = self.db.find({"location.state": {"$beginsWith": "New "}})
             self.assertEqual(len(docs), 2)
             self.assertUserIds([2, 10], docs)
```

[the remark this answers is not in the document]
Comment (willholley) [reply src/mango/src/mango_selector_text.erl]: thanks - I included the fix and added some covering test cases for escaped characters.

**Comment by willholley** [comment]
@pgj I think I've addressed all the comments, though will need to squash the commits

File: src/mango/test/03-operator-test.py [hunk 8]
```diff
@@ -141,6 +142,38 @@ def test_exists_false_returns_missing_but_not_null(self):
...
+                    docs = self.db.find(selector)
+                    self.assertEqual(len(docs), len(case["user_ids"]))
+                    self.assertUserIds(case["user_ids"], docs)
+
+        # non-string prefixes should return an error
+        def test_beginswith_invalid_prefix(self):
+            cases = [123, True, [], {}]
+            for prefix in cases:
+                with self.subTest(prefix=prefix):
+                    with self.assertRaises(HTTPError):
```
Comment (pgj) [inline src/mango/test/03-operator-test.py:174]: As far as I know, there is a different pattern in use throughout the test suite for testing cases like this one.  Advantages:
- There is no need to import `requests`, all the supporting code remains hidden in `mango`.
- The actual HTTP status code could be asserted.

Hence my recommendation:

```python
for prefix in cases:
    with self.subTest(prefix=prefix):
        try:
            self.db.find({"location.state": {"$beginsWith": prefix}})
        except Exception as e:
            self.assertEqual(e.response.status_code, 400)
        else:
            raise AssertionError("expected find error")
```

Curiously, you also use the same approach later in the change.

**Comment by pgj** [comment]
In the documentation, _1.3.6.1.1. Selectors Basics_ should also talk about `$beginsWith`?

https://github.com/apache/couchdb/blob/a5af77787bd58207bc63d822e9959103755325d8/src/docs/src/api/database/find.rst?plain=1#L201-L207

**Comment by pgj** [comment]
_1.3.6.1.8. Condition Operators_ seems to be missing `$beginsWith`:

https://github.com/apache/couchdb/blob/a5af77787bd58207bc63d822e9959103755325d8/src/docs/src/api/database/find.rst?plain=1#L676-L695

File: src/mango/test/03-operator-test.py (hunk 8 above)
Comment (willholley) [reply src/mango/test/03-operator-test.py:174]: I'm not too worried about importing requests but you are correct it would be better to test the explicit status_code here. In general I prefer the assertion functions provided by `unittest` because you get useful failure descriptions by default, but I don't think that's the case here.

File: src/docs/src/api/database/find.rst [hunk 9]
```diff
@@ -754,8 +760,10 @@ can itself be another operator with arguments of its own. This enables us to
 build up more complex selector expressions.
 
 However, only equality operators such as ``$eq``, ``$gt``, ``$gte``, ``$lt``,
-and ``$lte`` (but not ``$ne``) can be used as the basis of a query. You should
-include at least one of these in a selector.
+``$lte`` and ``$beginsWith`` (but not ``$ne``) can be used as the basis
```
[the remark this answers is not in the document]
Comment (willholley) [reply src/docs/src/api/database/find.rst:763]: I agree it's not an equality operator and we can change the language here.

File: src/mango/test/03-operator-test.py [hunk 10]
```diff
@@ -10,12 +10,13 @@
 # License for the specific language governing permissions and limitations under
 # the License.
 
+from requests.exceptions import HTTPError
```
Comment (pgj) [inline src/mango/test/03-operator-test.py:13]: This import is not needed any more.

**Comment by pgj** [comment]
+1

**Comment by big-r81** [comment]
@willholley
>Typically, CouchDB uses the [Review-Then-Commit](http://www.apache.org/foundation/glossary.html#ReviewThenCommit) (RTC) model of code collaboration, not _Commit-Then-Review_.

:stuck_out_tongue_winking_eye:

**Comment by pgj** [comment]
@big-r81 we have been in contact with @willholley about the PR out-of-band.  On merging the PR he was aware that I did not have any more pending comments on the contents.  I wanted to approve it formally but I was a bit late to the party hence I added the comment with the +1 to signal that I agree with the change.

**Comment by big-r81** [comment]
@pgj It was only a small friendly reminder 😉 btw, this reminds me of #4772...

**Comment by pgj** [comment]
Hrm...  Something is wrong with the merged version.  CI says the following:

```console
[2023-10-30T18:57:57.544Z] FAIL: test_basic (25-beginswith-test.BeginsWithOperator)
[2023-10-30T18:57:57.544Z] ----------------------------------------------------------------------
[2023-10-30T18:57:57.544Z] Traceback (most recent call last):
[2023-10-30T18:57:57.544Z]   File "/home/jenkins/workspace/kins-cm1_FullPlatformMatrix_main/centos7/build/src/mango/test/25-beginswith-test.py", line 49, in test_basic
[2023-10-30T18:57:57.544Z]     self.assertEqual(len(docs), 2)
[2023-10-30T18:57:57.544Z] AssertionError: 0 != 2
[2023-10-30T18:57:57.544Z] 
[2023-10-30T18:57:57.544Z] ======================================================================
[2023-10-30T18:57:57.544Z] FAIL: test_compound_key (25-beginswith-test.BeginsWithOperator)
[2023-10-30T18:57:57.544Z] ----------------------------------------------------------------------
[2023-10-30T18:57:57.544Z] Traceback (most recent call last):
[2023-10-30T18:57:57.544Z]   File "/home/jenkins/workspace/kins-cm1_FullPlatformMatrix_main/centos7/build/src/mango/test/25-beginswith-test.py", line 68, in test_compound_key
[2023-10-30T18:57:57.544Z]     self.assertEqual(len(docs), 1)
[2023-10-30T18:57:57.544Z] AssertionError: 0 != 1
```

**Comment by pgj** [comment]
Locally (with Dreyfus/Clouseau enabled) `make mango-test` works for me.

**Comment by nickva** [comment]
It probably a flaky test. We should investigate as we've been trying to squash some of those lately. Is the race condition fairly obvious? Usually we can add a retry loop.

**Comment by pgj** [comment]
It is not flaky, these errors happen every time -- but only on CentOS 7.  I was able to reproduce the issue locally with the help of a container created from the `apache/couchdbci-centos:7-erlang-24.3.4.10` Docker image.

**Comment by nickva** [comment]
> It is not flaky, these errors happen every time -- but only on CentOS 7.

Oh very interesting, definitely worth investigating. One thing that immediately jumps to mind is that CentOS 7 has an older libicu library which may have different collation rules. Any end/start (min/max) marker perhaps might collate differently there.

**Comment by willholley** [comment]
@big-r81 oops on the premature merge - I'm too used to working with repos with branch protection enabled in GH and didn't think to manually check for the +1.

**Comment by willholley** [comment]
curious that the failure didn't show in the PR checks - it definitely smells like a collation issue and likely an issue with the special characters used in the tests rather than the implementation of `$beginsWith`

**Comment by big-r81** [comment]
@willholley np, only a reminder. We will add this review-branch-protection in the future. At the moment, we will get the "green" merge button, if the CI runs through...

The failure didn't show up in the PR checks, because CentOS and other OSes are only called in the full CI run. For PRs, we only test a linux distro (debian) with different erlang version (24,25,26).

**Comment by big-r81** [comment]
@pgj can you test on your local CentOS vm the state before the PR?

**Comment by pgj** [comment]
Yeah, basically [the lack of support for `0xFFFF` in older ICU versions](https://www.unicode.org/reports/tr35/tr35-collation.html#tailored_noncharacter_weights) strikes again here, similar to what happened in #3491.  As the standard says:

_U+FFFF: This code point is tailored to have a primary weight higher than all other characters. This allows the reliable specification of a range, such as “Sch” ≤ X ≤ “Sch\uFFFF”, to include all strings starting with "sch" or equivalent._

Given that the other issue was fixed by unconditionally applying a shim on top of the ICU string comparison call, we should do the same here?  @nickva what do you think?

**Comment by nickva** [comment]
@pgj excellent find!

> Given that the other issue was fixed by unconditionally applying a shim on top of the ICU string comparison call, we should do the same here? @nickva what do you think?

It could work. The previous fix was easy enough as it was an isolated element -- a single marker. This one might involve doing some string wrangling in C, which is always fun, and by fun I mean dangerous, of course ;-)

I guess we could extend the current check to first see if the string ends in `\uFFFFF` and handle that special case.

**Comment by willholley** [comment]
Looking at https://icu.unicode.org/download/, it looks like there was a [rewrite of the collator](https://icu.unicode.org/design/collation/v2) for ICU 53 which includes better support for code points. For `$beginsWith` we would need to use a different unicode character for the upper bound I think, so there's a question around whether we make that conditional on the ICU version or not.

**Comment by willholley** [comment]
I wonder whether there's a pragmatically high code point we could use as the high bound in `$beginsWith` until we drop support for Centos 7 (it goes out of support next year)?

**Comment by nickva** [comment]
Another sentinel value might work, good idea.

CentOS is EOL next year so then we can deprecate support for it and remove a few hacks we have for it.

**Comment by pgj** [comment]
> Another sentinel value might work, good idea.

Sure but what should be that value specifically?  My impression is that 0xFFFF was created for that exact purpose because it was missing before.  It would be good to know how others have implemented this feature in lack of this symbol.

**Comment by nickva** [comment]
> Sure but what should be that value specifically? My impression is that 0xFFFF was created for that exact purpose because it was missing before. It would be good to know how others have implemented this feature in lack of this symbol.

Looking over at the PR itself, one issue that jumps out is in the line `<<Arg/binary, 16#10FFFF>>`. We don't actually build a binary with 3 extra bytes at the end: 10, FF and FF; instead we're appending an integer with the size of 8 bits which gets maxed out at FF.

```
io:format("~w~n", [<<"A",16#10FFFF>>]).
<<65,255>>
```

At some point the ICU library decided to treat that differently but they could have as well have thrown an error too.

But then, if even if we did `16#10FFFF:24` it would still hit the same special case of the API implementation. What we're passing in is a utf8 encoded binary but then append a raw unicode code point to it.

So then the fix might be to actually encode the `U+10FFF`:

```
io:format("~w~n", [<<16#10FFFF/utf8>>]).
<<244,143,191,191>>
```

Even better, since unicode standard seems to define a max sortable code point `U+FFFF`, we should use that instead of `U+10FFFF`.  So we could define a max as `unicode:characters_to_binary([16#FFFF])` or the shorter version of `<<16#FFFF/utf8>>`.

```
io:format("~w~n", [<<16#FFFF/utf8>>]).
<<239,191,191>>
```

Give `<<Arg/binary, 16#FFFF/utf8>>` a try, that should work on all the OS-es. It probably calls for extra unit tests too. 

One interesting case to think about is that with `/utf8` binary type we'll crash if the string is not a valid utf8 binary (say if somehow jiffy allowed through a surrogate pair prefix or we then building /utf8 binary should blow up).

**Comment by pgj** [comment]
Thanks @nickva!  I will experiment with the options you described above and get back to you with the results.

**Comment by willholley** [comment]
thanks @nickva - this solution looks good in my tests using the CentoOS container

**Comment by pgj** [comment]
> Even better, since unicode standard seems to define a max sortable code point `U+FFFF`, we should use that instead of `U+10FFFF`.

Erlang itself supports `U+10FFFF` and it is numerically greater than `U+FFFF`.

> Give `<<Arg/binary, 16#FFFF/utf8>>` a try, that should work on all the OS-es.

It works nicely.  Jenkins seems to be happy with that, see #4829 for the details.
